admin/ubuntu.py

```diff
@@ -0,0 +1,186 @@
+    "ubuntu_variant",
+    "release_cycle",
+    "release_date",
+)
+
+HEADINGS_DETAIL = HEADINGS_BASE + (
+    "ec2_image_type",
+    "architecture",
+    "region",
+    "ami_id",
+    "aki_id",
+    "unknown1",
+    "hypervisor",
+)
+
+UbuntuBase = new_record_type("UbuntuBase", HEADINGS_BASE)
+UbuntuDetail = new_record_type("UbuntuDetail", HEADINGS_DETAIL)
+
+HEADINGS_BY_TYPE = pmap({
+    UbuntuBase: HEADINGS_BASE,
+    UbuntuDetail: HEADINGS_DETAIL,
+})
+
+
+def pclass_issubset(subset, superset):
+    return set(
+        subset.serialize().items()
+    ).issubset(
+        set(superset.serialize().items())
+    )
+
+
+def filter_for_pclass(pclass_type, pclass_kwargs):
+    pclass_fields = pclass_type._pclass_fields
+    # Does the pclass support all the proposed kwargs
+    unexpected_keys = set(
+        pclass_kwargs.keys()
+    ).difference(
+        set(pclass_fields.keys())
+    )
+    if unexpected_keys:
+        raise ValueError(
+            "Unexpected keyword arguments for '{}'. "
+            "Found: {}. "
+            "Expected: {}.".format(
+                pclass_type.__name__,
+                list(unexpected_keys),
+                pclass_fields.keys(),
+            )
+        )
+
+    fields_to_validate = {
+        key: field
+        for key, field
+        in pclass_fields.items()
+        if key in pclass_kwargs
+    }
+    filter_type = type(
+        "Filter_{}".format(pclass_type.__name__),
+        (PClass,),
+        fields_to_validate
+    )
+    return filter_type(**pclass_kwargs)
+
+
+def url_records(url, record_type):
+    headings = HEADINGS_BY_TYPE[record_type]
+    for line in urlopen(url).readlines():
+        yield record_type(
+            **dict(zip(
+                headings,
+                line.decode('utf-8').rstrip("\n").split("\t"))
+            )
+        )
+
+
+def filter_records(records, search_record):
+    for record in records:
+        if pclass_issubset(
+                subset=search_record,
+                superset=record
+        ):
+            yield record
+
+
+def latest(**kwargs):
+    [latest] = filter_records(
+        records=url_records(
+            url_mangle(URL_LATEST_TEMPLATE, kwargs),
+            UbuntuBase,
+        ),
+        search_record=filter_for_pclass(
+            pclass_type=UbuntuBase,
+            pclass_kwargs=dict(
+                (k, kwargs[k])
+                for k in ("ubuntu_name", "ubuntu_variant", "release_cycle")
+            )
+        )
+    )
+    return filter_records(
+        records=url_records(
+            url_mangle(URL_DETAILS_TEMPLATE, kwargs),
+            UbuntuDetail,
+        ),
+        search_record=filter_for_pclass(
+            pclass_type=UbuntuDetail,
+            pclass_kwargs=freeze(kwargs).set(
+                "release_date", latest.release_date,
+            )
+        )
+    )
+
+
+class AMISearchUbuntuOptions(Options):
+    """
+    Options.
+    """
+    optParameters = [
+        ['release-cycle', None, u"daily",
```

Comment:
I chose daily as that's what I've always used when looking up the AMI IDs manually.

I'd been assuming that `released` meant that you get an Ubuntu image based on the packages at the time of first release, but that's not it:
 * https://cloud-images.ubuntu.com/query/trusty/server/released.current.txt

Were updated on 20160222. I'll stick with daily for now, but might change if I find a good explanation of the released list.